src/function/gds/output_writer.cpp

```diff
@@ -144,6 +148,18 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
     }
 }
 
+bool PathsOutputWriter::checkPathNode(const std::vector<ParentList*>& path) const {
+    if (pathNodeMask == nullptr) {
+        return true;
+    }
+    for (auto i = 0u; i < path.size() - 1; ++i) {
```

Comment:
The logic of applying the node predicate is not very simple. Adding a bit of comment here would help. What does path represent, the sequence of edges? Why are you looping  from 0 to path.size()-2 (inclusive), etc.